All nodes request helpers

gavin-norman-sociomantic opened the pull request.

Commit:
Add a usage example (documented unittest) for RequestCore

gavin-norman-sociomantic commented:
Based on #19 (based on #8). Blocked until parent PRs are merged.

gavin-norman-sociomantic commented:
Needs release notes.

gavin-norman-sociomantic commented:
Rebased on v4.x.x with #19 merged. Unblocked.

nemanja-boric-sociomantic commented:
Redish

gavin-norman-sociomantic commented:
Ha, that curious D1 error where you can shove template parameters on something that's not a template -- and the compiler ignores them!

nemanja-boric-sociomantic commented:
D1 helps you being more productive.

nemanja-boric-sociomantic commented on src/swarm/neo/client/mixins/RequestCore.d:
```diff
@@ -309,7 +309,7 @@ public template RequestCore ( RequestType request_type_, ubyte request_code,
     ***************************************************************************/
 
-    private static void notify ( ref UserSpecifiedParams params,
+    public static void notify ( ref UserSpecifiedParams params,
```
release notes?

nemanja-boric-sociomantic replied:
Documentation also says "Private helper"

nemanja-boric-sociomantic replied:
Documentation states "private helper", also perhaps missing release notes.

Commit:
Make notify function of RequestCore public

This enables it to be called from various helpers, which was previously
not possible.

gavin-norman-sociomantic commented:
Updated with release notes.

nemanja-boric-sociomantic commented on src/swarm/neo/client/mixins/AllNodesRequestCore.d:
```diff
@@ -0,0 +1,648 @@
+            }
+            finally
+            {
+                this.initialiser.reset();
```
Would it be a problem if the `run()` exited because `this.connector()` failed, so the `this.initialiser.initialise()` was not yet executed?

gavin-norman-sociomantic replied:
I don't think so. `initialiser.reset` should just reset the initialiser to its initial, default state. If `initialiser.inisialise` hasn't been called, it should be in its default state anyway.

Did you have anything specific in mind?

nemanja-boric-sociomantic replied:
Nope, that's all clarification I need. Thanks!

nemanja-boric-sociomantic commented on src/swarm/neo/client/mixins/AllNodesRequestCore.d:
```diff
@@ -0,0 +1,648 @@
+    RequestOnConn.EventDispatcherAllNodes conn )
+{
+    bool connected = false;
+    do // TODO: maybe don't need a loop, if there are no other resume codes to handle
```
This loop is not needed, indeed. The reason is that when `waitForReconnect` returns, we get either `connected = true`, or the program aborts (see `default` clause in switch).

gavin-norman-sociomantic replied:
Right, it was just an old note to myself that I forgot to address before pushing.

nemanja-boric-sociomantic commented on src/swarm/neo/client/mixins/AllNodesRequestCore.d:
```diff
@@ -0,0 +1,648 @@
+/*******************************************************************************
+
+    Struct template encapsulating the core logic for initialising a single
+    request-on-conn of an all-nodes request. Initialisation is deinfed as:
```
typo `deinfed`

nemanja-boric-sociomantic commented on src/swarm/neo/client/mixins/AllNodesRequestCore.d:
```diff
@@ -0,0 +1,648 @@
+    public bool initialise ( )
+    {
+        // establishConnection() should guarantee we're already connected
+        assert(this.conn.waitForReconnect() == 0);
```
I almost said "there shouldn't be any sideeffects inside `assert`, since it would be removed when compiling with `-release`", but this is essentially a way to say `assert(this.conn.isConnected())`. Maybe we should make that clearer (not in this PR).

david-eckardt-sociomantic replied:
:+1:
`waitForReconnect` seems to potentially have a side effect (that is, to wait), and IMHO assertions should not have any side effects regardless of whether code relies on them or not.

gavin-norman-sociomantic replied:
Ah, you're right. In the case where the assertion will _fail_, there could be side-effects. That's obviously not what I intended. Yeah, `assert(this.conn.isConnected())` is what I intended. I'll have a look at how to get that information.

gavin-norman-sociomantic replied:
Updated with an extra commit adding `client.RequestOnConn.AllNodesEventDispatcher.connection_status`.

nemanja-boric-sociomantic commented on src/swarm/neo/client/mixins/AllNodesRequestCore.d:
```diff
@@ -0,0 +1,648 @@
+
+    ***************************************************************************/
+
+    public uint initialising;
```
I wonder if we make all this field `private`, would that make compiler enforce that it's touched only by the other entities in this module. This would enforce that the user defined code doesn't change this field.

gavin-norman-sociomantic replied:
In this PR, it can be. It's read externally in a subsequent PR, but I'll change it to private here and add a getter in the other PR.

nemanja-boric-sociomantic replied:
:+1: I wasn't aware of the next PR.

nemanja-boric-sociomantic commented on src/swarm/neo/client/mixins/AllNodesRequestCore.d:
```diff
@@ -0,0 +1,648 @@
+
+    ***************************************************************************/
+
+    public void reset ( )
```
Nothing to do with the `all_nodes` in case of reset?

gavin-norman-sociomantic replied:
Nope. `AllNodesRequestSharedWorkingData.initialising` is reset in the `scope(exit)` of `initialise` and `AllNodesRequestSharedWorkingData.called_started_notifier` should never be reset because it must be called strictly once ever during the lifetime of the request.

gavin-norman-sociomantic replied:
I've added this explanation to the method doc.

Commit:
Add client RequestOnConn.AllNodesEventDispatcher.connection_status

gavin-norman-sociomantic commented:
Updated:
* Removed TODO / loop.
* Added connection status getter, use that in assert.
* Made `AllNodesRequestSharedWorkingData.initialising` private.
* Fixed typo.

Commit:
Add helpers encapsulating core logic of all-nodes request handling
